### PR TITLE
Open graph dynamic image generation

### PR DESCRIPTION
**Pull request by bmuenzenmeyer**
## Description

I've started dynamic open graph image generation.

This is still heavily WIP, and I need help landing it.

- [x] route location correct? I tried adding `route.tsx` elsewhere but couldn't figure it out. Vercel examples are way simpler than our structure.
- [x] how to add the OG images dyanmically into page metadata, I imagine via https://github.com/nodejs/nodejs.org/blob/main/next.dynamic.mjs#L192 
- [x] align on decisions within. 

This process uses documentation found within https://vercel.com/docs/functions/og-image-generation#getting-started and multiple techniques from https://vercel.com/docs/functions/og-image-generation/og-image-examples combined.

## Validation

http://localhost:3000/en/next-data/og?title=Announcing%20the%20new%20Node.js%20website&type=announcement

there are IMAGES! 😄 

image
image
image

### Check List

- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [x] I have run `npx turbo format` to ensure the code follows the style guide.
- [x] I have run `npx turbo test` to check if all tests are passing.
- [x] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

File: app/[locale]/og/route.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      : undefined;
+
+    //?type=<type>
+    const type = searchParams.get('type') ?? 'announcement';
+
+    const typeAttributes: { [key: string]: string } = {
+      announcement: tailwindConfig.theme.colors.green['700'],
+      release: tailwindConfig.theme.colors.info['600'],
+      vulnerability: tailwindConfig.theme.colors.warning['600'],
+    };
```
Comment (bmuenzenmeyer) [inline app/[locale]/og/route.tsx]: we can re-use our TW config here - staying as dry as possible...

File: app/[locale]/og/route.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    return new ImageResponse(
+      (
+        <div
+          tw={`relative flex items-center justify-center bg-black
```
Comment (bmuenzenmeyer) [inline app/[locale]/og/route.tsx]: note - this OG tailwind technique cannot use our styles, it must use the static `tw` attribute.
That means I couldn't simply re-use `Preview`

**Review by bmuenzenmeyer — COMMENTED** [review]
self-review

File: app/[locale]/og/route.tsx
Comment (ovflowd) [inline app/[locale]/og/route.tsx:1]: Can you put it under the next-data folder? `next-data/og`?

**Comment by ovflowd** [comment]
What do you need help with to land this?

**Comment by bmuenzenmeyer** [comment]
> What do you need help with to land this?

my main concerns were

- ✅ ~where to put it in our filesystem~
- ✅ ~how to add it to each blog (as I linked, I think that would work)~

I'd be happy with commits added to it

edit: also happy to do it - just needed some corroborated direction

**Comment by bmuenzenmeyer** [comment]
this now adds the following to every page

image

here's a comparison of post `Preview` and the `og` image

> this isn't exact.  I tried using `bg-neutral-950` as the tw base and it broke. I might play with it, but welcome any TW-fu



<br class="Apple-interchange-newline">
</body>
</html>

image


do we want to constrict this to only blog posts? right now it's on every page

even regular pages

image

**Comment by bmuenzenmeyer** [comment]
okay - this is good enough that I'd love a more thorough poke at it

**Comment by bmuenzenmeyer** [comment]
an OG API concern I'd like specific feedback on - do we need to encrypt our parameters like https://vercel.com/docs/functions/og-image-generation/og-image-examples#encrypting-parameters suggests? NOT doing so means someone else could use the API to spoof our official previews, right? do we care?

**Comment by ovflowd** [comment]
> an OG API concern I'd like specific feedback on - do we need to encrypt our parameters like [vercel.com/docs/functions/og-image-generation/og-image-examples#encrypting-parameters](https://vercel.com/docs/functions/og-image-generation/og-image-examples#encrypting-parameters) suggests? NOT doing so means someone else could use the API to spoof our official previews, right? do we care?

I don't see any harm, but if you feel someone could misuse this in a way that affects us poorly, sure.

**Comment by ovflowd** [comment]
FYI, this branch needs to be rebased

**Comment by bmuenzenmeyer** [comment]
@ovflowd i need guidance on if any of this is still feasible with static export 

see https://github.com/nodejs/nodejs.org/actions/runs/8111322926/job/22170364122?pr=6357#step:10:34 - next wants me to statically generate every param - i suppose we have the blog data so it's possible???, but really starts to fight the framework for something that should be dynamic

**Comment by ovflowd** [comment]
> @ovflowd i need guidance on if any of this is still feasible with static export
> 
> see [nodejs/nodejs.org/actions/runs/8111322926/job/22170364122?pr=6357#step:10:34](https://github.com/nodejs/nodejs.org/actions/runs/8111322926/job/22170364122?pr=6357#step:10:34) - next wants me to statically generate every param - i suppose we have the blog data so it's possible???, but really starts to fight the framework for something that should be dynamic

No, `generateStaticParams` only means if you want to generate something statically. You don't need to provide any route. Let me make amends to this PR so that it works with the intended approach.

**Review by ovflowd — APPROVED** [review]

**Comment by ovflowd** [comment]
@bmuenzenmeyer can you make a hotfix?

**Comment by bmuenzenmeyer** [comment]
hotfixed via https://github.com/nodejs/nodejs.org/commit/604b4b471bbf3355265572b4b621afb24c1ed194

**Pull request merged**